### ISSUE TITLE
Working with Redux Reducers

### ISSUE THREAD

**Issue by xareelee** [facebook/flow#3757]
### TL;DR

I need to know how to use a key type to match a type in the disjoint types.

### Description

My **Redux actions** have two kinds of types: w/ `payload` or w/o `payload`.

```js
type ActionName = 
  | 'LOGIN'
  | 'LOGOUT'
  ;

// action without payload
type ActionT<A: string> = {|
  type: A
|};

// action with payload
type ActionP<A: string, P> = {|
  type: A,
  payload: P
|};

type Action = 
  | ActionP<'LOGIN', UserInfo>
  | ActionT<'LOGOUT'>
  ;
```

Now I use a function `buildReducer` to build a reducer for types:

```js
const buildReducer = (initialState, reducerMap) =>
  (state = initialState, action) => {
    const reducer = reducerMap[action.type];
    return (reducer) ? reducer(state, action) : state;
  };

const initialState: {
  userInfo: UserInfo
} = {};
const reducers = {};

reducers['LOGIN'] = (state, action) =>
  ({ ...state, userInfo: action.payload });

reducers['LOGOUT'] = (state, action) =>
  ({ ...state, userInfo: null });

const combinedReducer = buildReducer(initialState, reducers);
```

#### Question 1

I want to assign a type to `reducers`, and make flowtype check the following syntax with the wrong types:

```js
reducers['LOGOUT'] = (state, action) =>
  ({ ...state, userInfo: action.payload });  // Error: Action `LOGOUT` doesn't have a payload.
```

How do I do to check the action with the key `'LOGOUT'` in the `reducers`?

#### Question 2

When I assigned the `reducers` a type, it raise errors:

```js
type State = any;
type Reducer = (State, Action) => State

const reducers: { [ActionName]: Reducer} = {};

reducers['LOGOUT'] = (state, action) =>
  ({ ...state, userInfo: action.payload });  // Will raise an error with flowtype:
```

```
  ({ ...state, userInfo: action.payload });
                                ^^^^^^^ property `payload`. Property not found in
  type ActionT<A: string> = {|
                            ^ object type.

```

I want exact disjoint types. How do I fix that?

#### Question 3

How can I get an exact type from the disjoint types with a key? For example:

```js
// ActionGetType is a type (pseudo code) which accepts a generic type and return a type from the disjoint types `Action`.
type ActionLogin = ActionGetType<'Login'>; // equals to `ActionP<'LOGIN', UserInfo>` or `{ type:  'LOGIN', payload: UserInfo }`
```

**Comment by madbence** [comment]
to answer your third question, it's kinda possible:

```js
type ActionTypeMap = {
  LOGIN: ActionP<'LOGIN', UserInfo>,
  LOGOUT: ActionT<'LOGOUT'>,
};

type LoginType = $PropertyType<ActionTypeMap, 'LOGIN'>;
({ type: 'LOGIN', payload: {}}: LoginType)
```

unfortunately, this looks a bit ugly, because `$PropertyType` accepts only string literals, otherwise you could use something like `type GetActionType<T> = $PropertyType<ActionTypeMap, T>`

the other issues are a bit harder, and i'm not sure that it's possible to fix them without additional code (or unsafe casting). the main problem is that the type of `reducerMap[action.type]` is not refined, it is `(State, Action) => State`, and `Action` can represent both `LOGIN` and `LOGOUT`. you can cast the action to `any`, and then you can annotate your reducers individually:

```js
type State = {
  userInfo?: UserInfo,
};

const buildReducer = (initialState, reducerMap) =>
  (state: State = initialState, action: Action) => {
    const reducer = reducerMap[action.type];
    return reducer ? reducer(state, (action: any)) : state;
  };

const initialState: State = {};
const reducers = {};

reducers['LOGIN'] = (state, action: ActionP<'LOGIN', UserInfo>) =>
  ({ ...state, userInfo: action.payload });

reducers['LOGOUT'] = (state, action: ActionT<'LOGOUT'>) =>
  ({ ...state, userInfo: undefined });

const combinedReducer = buildReducer(initialState, reducers);
```

or you can refine the type of `action` manually (i guess `buildReducer` was created in the first place to replace a big `switch` statement, so this is not really useful)

[try flow](https://flow.org/try/#0C4TwDgpgBAqgzhATgSQHYDMD2UC8UDeAvgNwBQpokUAggMbACWmqAcgIYC20epUUAPlADkAGQDyAcWQshvAcPESxMACqy+ZCuGh1GzFQB5qALihxgiBqgDmAPlwF+cyhFPVS-EuRc16TVAAKRqbmljYANFAB9nj4TnwubuFyYGwgADaYbAAmpgEeXlpUuv4OcoIlzEGiktJCkfBIaFi25b56qIY1SqpCrRre2lAAysBswNwEcgCuCCgYmAD8po3zWMmFPpWd2gCybGAO+HKK0m5+VQbddQ1zzZi2yXyKyirnHV0vvY+km0MimGsVhUQzwABIAohMJBEKAQZAjBcdpB9mBItcZLZNAAKfBQRIKWoySKpDJZXIEQiEUwAoHIiAASnItGY5igACNpgx0tkAEoQbLTWhIBzYqwMRhsdKjcYQSKIAVCpCohm4fpQbHmWWmGUTBziyXSsYTSJsJHvfyqnD2Y58Pgs1BshWC4WIBzOpWIVEAbTNHQAdC4ALpkO1QBXAaaIVDhxWuqCLWMupCa41yjV+-ymNioEAM1UhNOhqCFB1sg0MKW61wjNNHLxl4BJz1weuaD2uuDejFCIOirUmqCZ5gWy4925NBa2K3q3FQf0Lgfp2aTrDZpH+0mZHIlhntuNILvdV69-tp03m9r+T6SE-TtVyOcL-1LyIrtaYUzTVDZCDoKwCrumiNlALIcOyAF8gebp4Jy3JQcmiBiqgEqVkasrytBcB7kAA)

**Comment by xareelee** [comment]
@madbence Got it. Thanks for your answer.

It seems that flowtype needs to add some features to support Redux deeply. 😃
